### DISCUSSION TITLE
Stuck with the calculation of Solution 2 Compass

### DISCUSSION THREAD

**Discussion by NitinSaxenait** [japaric/discovery#383] [Q&A]
Hi Everyone, I am working on the stm32-discovery board for the last 4 months and successfully completed 13 chapters from the book. Currently, I am working on the chapter [LED compass](https://docs.rust-embedded.org/discovery/15-led-compass/index.html) and I have passed [Take 1](https://docs.rust-embedded.org/discovery/15-led-compass/take-1.html) of the chapter using lsm303agr instead of lsm303dlhc.
Now I am working with [Take 2](https://docs.rust-embedded.org/discovery/15-led-compass/take-2.html) and got to know that we are going to find out the precise angle of magnetic field using atan2 function but I am really not getting the calculation which we are using inside the if-else block.
`let theta = (y as f32).atan2(x as f32); // in radians

        let dir = if theta < -7. * PI / 8. {
            Direction::North
        } else if theta < -5. * PI / 8. {
            Direction::Northwest
        } else if theta < -3. * PI / 8. {
            Direction::West
        } else if theta < -PI / 8. {
            Direction::Southwest
        } else if theta < PI / 8. {
            Direction::South
        } else if theta < 3. * PI / 8. {
            Direction::Southeast
        } else if theta < 5. * PI / 8. {
            Direction::East
        } else if theta < 7. * PI / 8. {
            Direction::Northeast
        } else {
            Direction::North
        };`
Can someone please help me with this. Thanks

**Comment by adamgreig** [comment]
The compass tells us the strength of the magnetic field in an `x` and `y` direction, but we want to turn this into a cardinal direction (N, NW, W, SW, S, etc). First we turn the (x, y) reading into an angle (a bearing) using `atan2(y, x)`, which computes the inverse tan: there's [more information on Wikipedia](https://en.wikipedia.org/wiki/Atan2) including some diagrams:

<a href="https://en.wikipedia.org/wiki/File:Atan2definition.svg"><img src="https://upload.wikimedia.org/wikipedia/commons/a/ad/Atan2definition.svg" width=500></a>

Then once we have the angle (in units of radians, so instead of 0-360° it goes from -pi up to +pi), we use this if/else if/else code to turn the angle into a direction. The `atan2` method, when called with these magnetic field strengths, returns something near 0 when the field is south, and something near +-pi when the field is north, so we're just splitting the angles up into different segments. In fact this means our bearings are backwards (normally 0° would be north and 180° south, but here 0° is south and 180° is north), but it doesn't really matter since we just use the right numbers in the code to get the right direction.

Something like this diagram from Wikipedia, but label the (1, 0) point as South and the (-1, 0) point as North:

<a href="https://en.wikipedia.org/wiki/File:Unit_circle_angles_color.svg"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4c/Unit_circle_angles_color.svg/600px-Unit_circle_angles_color.svg.png" width=500></a>

Does that help?

**Comment by NitinSaxenait** [comment]
@adamgreig This was really helpful and after this, I did some research also and found out that PI/8 contains the angle from `0 to 0.3925 under which comes the South. Similarly  0.3925×3 => 1.1775(SE), 0.3925×5 => 1.9625(E) & 0.3925×7 => 2.7475(NE)` and board actually change the led at point(value) differences. Amazing..!!  
Okay so now I know now we are dividing -pi to +pi in 8 parts but I am still confused about how we calculating these values using x3, x5, and x7. Can you please teach me more about it. This is really interesting. Thanks
